File: benchmark/prisoners/prisoner.js
const _ = require('lodash');
const { performance } = require('perf_hooks'); // ใช้สำหรับวัดเวลาที่แม่นยำ

const numPlays = 10000; // แนะนำให้ใช้หมื่นรอบก่อน เพราะแบบ Random จะช้ามาก
const numPrisoners = 100;

const setupSecrets = () => {
    let secrets = [];
    for (let i = 0; i < numPrisoners; i++) {
        secrets.push(i);
    }
    return _.shuffle(secrets);
}

const playOptimal = () => {
    let secrets = setupSecrets();
    for (let p = 0; p < numPrisoners; p++) {
        let success = false;
        let choice = p;
        for (let i = 0; i < numPrisoners / 2; i++) {
            if (secrets[choice] === p) {
                success = true;
                break;
            }
            choice = secrets[choice];
        }
        if (!success) return false;
    }
    return true;
}

const playRandom = () => {
    let secrets = setupSecrets();
    for (let p = 0; p < numPrisoners; p++) {
        // สร้างรายการสุ่มเลือก 50 ใบจาก 100 ใบ
        let choices = _.sampleSize(_.range(numPrisoners), numPrisoners / 2);
        let success = false;
        for (let i = 0; i < choices.length; i++) {
            if (secrets[choices[i]] === p) {
                success = true;
                break;
            }
        }
        if (!success) return false;
    }
    return true;
}

// ฟังก์ชันหลักสำหรับทำ Benchmark
const runBenchmark = (name, fn) => {
    console.log(`Running ${name}...`);
    const startTime = performance.now(); // เริ่มจับเวลา
    let successCount = 0;

    for (let i = 0; i < numPlays; i++) {
        if (fn()) successCount++;
    }

    const endTime = performance.now(); // สิ้นสุดจับเวลา
    const durationSeconds = ((endTime - startTime) / 1000).toFixed(3);
    const winRate = ((successCount / numPlays) * 100).toFixed(2);

    return { name, winRate, durationSeconds };
}

// --- ส่วนแสดงผล ---
console.log(`--- 100 Prisoners Benchmark (${numPlays.toLocaleString()} rounds) ---`);

const results = [
    runBenchmark("Optimal Strategy", playOptimal),
    runBenchmark("Random Strategy", playRandom)
];

console.table(results); // แสดงผลเป็นตารางสวยงาม

File: benchmark/prisoners/prisoner.py
import random
import time

def play_random(n, num_prisoners=100):
    pardoned = 0
    sampler = list(range(num_prisoners))
    for _ in range(n):
        in_drawer = list(range(num_prisoners))
        random.shuffle(in_drawer)
        success = True
        for prisoner in range(num_prisoners):
            # สุ่มเลือก 50 ใบ
            revealed = random.sample(sampler, num_prisoners // 2)
            if prisoner not in [in_drawer[i] for i in revealed]:
                success = False
                break
        if success:
            pardoned += 1
    return (pardoned / n) * 100

def play_optimal(n, num_prisoners=100):
    pardoned = 0
    in_drawer = list(range(num_prisoners))
    for _ in range(n):
        random.shuffle(in_drawer)
        success = True
        for prisoner in range(num_prisoners):
            reveal = prisoner
            found = False
            for go in range(num_prisoners // 2):
                card = in_drawer[reveal]
                if card == prisoner:
                    found = True
                    break
                reveal = card
            if not found:
                success = False
                break
        if success:
            pardoned += 1
    return (pardoned / n) * 100

def run_benchmark(n):
    results = []
    
    # ทดสอบ Optimal
    start = time.perf_counter()
    win_opt = play_optimal(n)
    time_opt = time.perf_counter() - start
    results.append(["Optimal Strategy", f"{win_opt:6.2f}%", f"{time_opt:8.4f}s"])

    # ทดสอบ Random
    start = time.perf_counter()
    win_rand = play_random(n)
    time_rand = time.perf_counter() - start
    results.append(["Random Strategy", f"{win_rand:6.2f}%", f"{time_rand:8.4f}s"])

    # พิมพ์ตารางแบบ Manual (คล้าย console.table)
    header = ["Strategy", "Win Rate", "Time Used"]
    print(f"\nSimulation count: {n:,}")
    print("-" * 45)
    print(f"{header[0]:<20} | {header[1]:<10} | {header[2]:<10}")
    print("-" * 45)
    for row in results:
        print(f"{row[0]:<20} | {row[1]:<10} | {row[2]:<10}")
    print("-" * 45)

if __name__ == '__main__':
    # แนะนำให้ใช้ 10,000 รอบก่อน เพราะแบบ Random ใน Python จะช้ากว่า JS พอสมควรครับ
    run_benchmark(10000)